services/src/main/java/org/keycloak/protocol/oidc/grants/JWTBearerAssertionGrantType.java

```diff
@@ -0,0 +1,341 @@
+
+import jakarta.ws.rs.core.MediaType;
+import jakarta.ws.rs.core.Response;
+import org.keycloak.OAuth2Constants;
+import org.keycloak.OAuthErrorException;
+import org.keycloak.TokenVerifier;
+import org.keycloak.common.VerificationException;
+import org.keycloak.events.Details;
+import org.keycloak.events.Errors;
+import org.keycloak.events.EventType;
+import org.keycloak.models.*;
+import org.keycloak.protocol.oidc.*;
+import org.keycloak.protocol.oidc.TokenManager;
+import org.keycloak.protocol.oidc.utils.AssertionGrantUtils;
+import org.keycloak.representations.AccessTokenResponse;
+import org.keycloak.representations.JsonWebToken;
+import org.keycloak.services.CorsErrorResponseException;
+import org.keycloak.services.Urls;
+import org.keycloak.services.managers.AuthenticationManager;
+import org.keycloak.services.managers.AuthenticationSessionManager;
+import org.keycloak.services.resources.admin.permissions.AdminPermissions;
+import org.keycloak.sessions.AuthenticationSessionModel;
+import org.keycloak.sessions.RootAuthenticationSessionModel;
+import org.keycloak.util.JsonSerialization;
+import org.keycloak.util.TokenUtil;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * OAuth 2.0 JWT Bearer grant type (RFC 7523)
+ *
+ * @author Ben Cresitello-Dittmar
+ */
+public class JWTBearerAssertionGrantType extends OAuth2GrantTypeBase {
+    private List<CrossDomainTrust> trustedIssuers;
+
+    @Override
+    public Response process(Context context) {
+        setContext(context);
+
+        event.detail(Details.AUTH_METHOD, "jwt_bearer");
+        event.client(client);
+
+        this.trustedIssuers = AssertionGrantUtils.getTrustedIssuerConfigs(realm, client);
+
+        checkAssertionGrantEnabled();
+        checkTrustedIssuerConfig();
+        checkConfidentialClient();
+
+        return authenticateWithAssertion();
+    }
+
+    /**
+     * Authenticate a user with a JWT bearer assertion
+     *
+     * @return The token endpoint response
+     */
+    protected Response authenticateWithAssertion() {
+        String assertion = formParams.getFirst(OAuth2Constants.ASSERTION);
+        checkAssertion(assertion);
+
+        // validate assertion sig
+        JsonWebToken jwt = validateAssertion(assertion);
+
+        UserModel requestedUser = getUser(jwt);
+        checkImpersonationAllowed(requestedUser);
+
+        // log user in (transient session)
+        UserSessionModel userSession = session.sessions().createUserSession(null, realm, requestedUser, requestedUser.getUsername(), clientConnection.getRemoteAddr(), "impersonate", false, null, null, UserSessionModel.SessionPersistenceState.TRANSIENT);
+        event.session(userSession);
+
+        setAssertionInUserSessionNotes(jwt, userSession);
+
+        AuthenticationSessionModel authSession = createAuthSession(requestedUser);
+        updateUserSessionFromClientAuth(userSession, clientAuthAttributes);
```

Comment:
Hi Ben,

It looks like the implementation on line 330 could be replaced with a call to the base class method `updateUserSessionFromClientAuth(userSession)`. This might allow us to remove the redundant code.

What do you think?

Thanks!